### ISSUE TITLE
Potential memory leaks due to NSDeallocObserver

### ISSUE THREAD

**Issue by caffeinepills** [mrJean1/PyCocoa#4]
I've noticed memory leakage with the `NSDeallocObserver` setup. Some objects just don't seem to get a dealloc call ever when hard referenced. (Unknown reason why.)

It seems this is a known possibility as the `ObjCInstance` has a kwarg for enable or disabling cache. However, there is no easy way to do this for a returned type. (Atleast from searching the repository).

Example:
```py
import pycocoa

NSDate = pycocoa.ObjCClass('NSDate')

while 1:
    pool = pycocoa.NSAutoreleasePool.alloc().init()
    
    date = NSDate.dateWithTimeIntervalSinceNow(0.0)  # doesn't dealloc

    pool.drain()  # dealloced fine

    del pool

    print(len(pycocoa.ObjCInstance._objc_cache))  # Infinite growth.
```

Not sure if this is an issue with Cocoa, the observer implementation, or both.

**Comment by mrJean1** [comment]
Thank you for reporting the issue with the example.

It turns out that `NSAutoreleasePool.drain` does _not invoke_ the `dealloc` method of the objects in the pool.  And as a result, those objects are not removed from the `_obj_cache`.

Plus, there doesn't seem to be a way to iterate over the objects in an `NSAutoreleasePool`.

Therefore, the only resolution is to  (a) clear the entire `_obj_cache` when an `NSAutoreleasePool` is `dealloc`ed and `drain`ed or (b) do not `_obj_cache` any objects at all.

**Comment by caffeinepills** [comment]
I think a) may not be viable due to the fact you are supposed to be able to nest `NSAutoreleasePool`s pools, so clearing out all objects would not work. b) is interesting, but may require work arounds for existing code, as I'm not sure what would happen if nothing gets cached.

**Comment by mrJean1** [comment]
(a) will only empty the `_objc_cache`, nothing else.  The objects originally cached remain alive until `dealloc`ed explicitly or `drain`ed if in a pool.  Btw, `pool.drain()` also `dealloc`s the pool itself.

**Comment by caffeinepills** [comment]
Let's say you have 2 pools (a and b) and objects created outside of a pool. If you create an instance outside of the pool, or in pool B, if pool A calls for the clearing of `_objc_cache` it will affect everything. Or is that not something to worry about?

**Comment by mrJean1** [comment]
Yes, draining any pool with clear the entire `_objc_cache` because there is only one `_obj_cache`.  Only the ObjC objects in pool A will be destroyed, along with their Python wrappers.  All other Python and ObjC objects are unaffected and pool B remains intact.

No, the `_objc_cache` is only intended to speed up ObjC instantiation.  Clearing the `_objc_cache` does not delete or destroy any Python or ObjC object.

To see the effect, change the keyword argument `cached=False` of `ObjCInstance.__new__` in module `pycocoa.runtime`.  Except for speed and the memory leak, the behavior should be the same.

**Comment by caffeinepills** [comment]
Running the example https://github.com/mrJean1/PyCocoa/blob/564880225245635e17bbd5e9480f5a3e40bb4756/test/simple_drawing.py

With `cached=False` produces errors.

**Comment by mrJean1** [comment]
Thank you for testing.  Using (b) `cached=False` is not feasible.  A new release based (a)  is forthcoming once all tests pass.

**Comment by mrJean1** [comment]
PyCocoa `23.1.18` is available on [PyPI](https://pypi.org/project/PyCocoa/23.1.18/) and [GitHub](https://github.com/mrJean1/PyCocoa) containing the (a) fix for the memory leak plus a few other, minor changes.

The tests pass, but please double check.  And close the issue if all is well.

**Comment by caffeinepills** [comment]
I tested the changes, and so far so good. The only minor issue I have seen so far is that if you are doing an `NSAutoreleasePool` and drain within a `ObjCSubclass` (for instance in the `simple_drawing` example's `drawRect`) when closing the Window (via X button), you get the same kind of error as above:

```py
Traceback (most recent call last):
  File "/Users/admin/PycharmProjects/venv/lib/python3.11/site-packages/pycocoa/runtime.py", line 1009, in objc_method
    return _pyresult(m(_pyself, *_pyargs(codes3, args)))
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/admin/PycharmProjects/no_cache.py", line 80, in windowWillClose_
    self.app.terminate_(self)
    ^^^^^^^^
  File "/Users/admin/PycharmProjects/venv/lib/python3.11/site-packages/pycocoa/runtime.py", line 634, in __getattr__
    raise AttributeError('no %r [class]method or property: %s' % (name, self))
AttributeError: no 'app' [class]method or property: _Delegate(<Id_t at 0x10f0c7c80>) of 0x6000009acb30
```

However, testing the drain during operation, I can confirm it doesn't leak anymore. 👍

**Comment by mrJean1** [comment]
None of Python 2.7.18, 3.6.5, 3.7.6, 3.8.10, 3.9.6, 3.10.8 and 3.11.1 reproduces the problem (on macOS 13.1).  In retrospect, that example is a bit suspicious in the `initWithFrame_` method, overwriting `self` in the very first line.

Attached is a version where `self` is replaced with `objc` in the first and last line of that method.  See if that gets rid the exception.

[simple_drawing23.01.19.zip](https://github.com/mrJean1/PyCocoa/files/10460158/simple_drawing23.01.19.zip)

**Comment by caffeinepills** [comment]
I tried that example and it still does it for me (I did have to comment out `import run as _`)

Here is my test I am doing to check (Add the imports for `NSDate` and `NSAutoreleasePool`): 
```py
    @_View.method('v@')
    def drawRect_(self, rect):

        # w, h = self.boundsSize().width, -.height
        b = rect or self.bounds()
        w = b.size.width * 0.5
        h = b.size.height * 0.5

        def _x(t, w):
            return (sin(t) + 1.) * w

        def _y(t, h):
            return (cos(t) + 1.) * h

        NSColor.whiteColor().set()
        NSRectFill(b)  # not a class

        NSColor.blackColor().set()
        for f in self.loop:
            p1 = NSPoint_t(_x(f, w), _y(f, h))
            for g in self.loop:
                p2 = NSPoint_t(_x(g, w), _y(g, h))
                NSBezierPath.strokeLineFromPoint_toPoint_(p1, p2)

        pool = NSAutoreleasePool.alloc().init()

        date = NSDate.dateWithTimeIntervalSinceNow_(0.0)

        pool.drain()
```
Just running this as a test exhibits the issue. If you aren't seeing the same issue, then maybe something with my environment setup.

**Comment by mrJean1** [comment]
Thank you.  That does reproduce the problem.

**Comment by caffeinepills** [comment]
The only issue I ran into with (c) is there no way to easily expose that for a return type.

I have thought of two possibilities as well and implemented (d) on another project as a quick fix. Neither solutions are perfect.
d) add `no_cached_return` to `ObjCBoundMethod`, for the user to specify as not cachable. So for instance `NSDate.dateWithTimeIntervalSinceNow_.no_cached_return()`, this sets a property to make the return `ObjCInstance `return as `cache=False` . Or the inverse `cache_return`, where by default everything is not cached, but you can specify something to be cached. Downside is, (in either solution) is user has to be aware of what needs to be cached and not cached.

e) Create a Python specific version of Autopool. This keeps track of the objects in it. We are using this as a more Pythonic way of a pool:
```py
class AutoReleasePool:
    def __enter__(self):
        self.pool = NSAutoreleasePool.alloc().init()
        return self.pool

    def __exit__(self, exc_type, exc_value, traceback):
        self.pool.drain()
        del self.pool
 ```
 
It's use would be:
 ```py
 with AutoReleasePool():
       dostuff()
 ```

This could be expanded in such a way:
```py
class AutoReleasePool:
    def __init__(self):
        self._pool_cache = {}
        self._old_cache = None

    def __enter__(self):
        self._old_cache = ObjCInstance._objc_cache  # store old cache
        ObjCInstance._objc_cache = self._pool_cache  # change cache to pool cache
        self.pool = NSAutoreleasePool.alloc().init()
        return self.pool

    def __exit__(self, exc_type, exc_value, traceback):
        self.pool.drain()
        del self.pool
        ObjCInstance._objc_cache = self._old_cache  # restore old cache
        self._pool_cache.clear()  # clear pool cache
 ```

Downside is, they can't really use the objc implementation without managing the cache themselves and would have to be specified this is the recommended way to use the pool.  Plus it may seem somewhat of a workaround? I'm not sure.

**Comment by mrJean1** [comment]
Both great suggestions e) was going to be the next attempt.  More forthcoming.

**Comment by caffeinepills** [comment]
Thanks for tracking this down. That does appear to have fixed the issue so far.

**Comment by mrJean1** [comment]
One enhancement should be to keep track of the (nested) pool cached objects are held in.  Then, only clear those objects from the `_objc_cache` which are held in the pool being drained.

**Comment by caffeinepills** [comment]
I will say there may be another memory leak somewhere, from what I can tell it is not observer/instance related. Should I create a separate issue for it?

**Comment by caffeinepills** [comment]
Closing as this seems to be resolved with #6
